### DISCUSSION TITLE
deduplication on shapshot receival

### DISCUSSION THREAD

**Discussion by slavanap** [openzfs/zfs#15066] [General]
I have a question regarding commands parameters, because it's not clear from the documentation.

I have a huge pool build up with backup snapshots for years. It has one encrypted dataset, root is not encrypted.

The pool is stored on a separate disk, no raid or anything, I utilize ZFS snapshots and compression capabilities for backup purposes only.

I want to rewrite the entire pool and dedup its contents with support of dedup special dev on SSD.

Questions:
1. How to send first snapshot of encrypted dataset to a new pool, preserving the encryption and dedup and recompress its data on the fly. (so -w option is not feasible; what are the options?)
2. How to dedup and recompress each of snapshots I send over after the initial one (presumably with -I option)? Is it guaranteed that dedup and recompression will happen automatically on an unencrypted dataset. 
3. If I change (increase) recordsize to 1M, will this option be regarded on receive and blocks that originally had recordsize=128K will be correctly concatenated?

Answers:

1. ```zfs recv -o encryption=aes-256-gcm -o keyformat=passphrase -o keylocation=file:///testfile -o dedup=verify -o recordsize=1M -o compression=zstd-10 -o atime=off -o relatime=on main/test2```
then
2. ```sudo zfs send --large-block -I main/test1@snap{,2} | sudo zfs recv -F main/test2```

3. do not change recordsize on test1 because if will reduce de-duplication effect.

**Comment by amotin** [comment]
If you are using encrypted replication, then you can't change anything, receive side blocks will be exactly the same as source, may only participate in dedup, since it works on different layer.  If replication is compressed but not encrypted, receiving side may have different indirect block sizes and few other things, but block will be the same, including original compression.  If replication is neither compressed nor encrypted, then receiving side may have different compression settings, but block sizes should remain the same unless you are looking for troubles.

**Reply by slavanap** [reply]
But I can decrypt the stream with no usage of -w option for zfs send.
Is there a way to uncompress it for receiving side as well so then receiving side could recompress it with a better algorithm?

**Reply by amotin** [reply]
If you use neither -w nor -c, then stream will be neither compressed nor encrypted.  In such case receiving side can use different algorithm.

**Reply by slavanap** [reply]
Thanks for that. This information is useful. The remaining bit is how to start "recreation" of an encrypted dataset with re-compression and de-duplication. After new dataset is created with an encryption key, ZFS does not allow to write first snapshot unencryptedly to it:
```
$ zfs send main/test1@snap | zfs recv main/test2
cannot receive new filesystem stream: destination 'main/test2' exists
must specify -F to overwrite it
```

```
# zfs send main/test1@snap | zfs recv -F main/test2
cannot receive new filesystem stream: zfs receive -F cannot be used to destroy an encrypted filesystem or overwrite an unencrypted one with an encrypted one
```

The only way I know how to copy encryption key is to use `-w` option, which also copies all data without a change of compression nor deduplication. So the remaining question is how to copy encryption key on a new dataset while sending first snapshot in an unencrypted way.

**Reply by amotin** [reply]
What if you just delete the destination first?  I have vague memories that there were problems with unencrypted replication of encrypted datasets.  IIRC last time I've looked on it it looked like a bug or artificial limitation to me, but I don't remember whether it was fixed or not.

**Reply by slavanap** [reply]
Well, if I remove it then the command above will create an unencrypted dataset

**Reply by amotin** [reply]
Ah, so you want unencrypted to replication to encrypted dataset.  I don't think it is impossible in theory, but I have doubts it works now.  Destination in that case would have its own keys, but the question is how to supply them to receive command.

**Reply by rincebrain** [reply]
You can happily do recv -o keyformat=xxx -o keylocation=xx if you like.

Note that you can't mix -p with sending without encryption, so you'd have to manually reproduce the properties you want on first recv as well.

Of course, it might panic your system like any encrypted send/recv, but that's not related.

**Reply by slavanap** [reply]
@rincebrain thanks, this one worked flawlessly
`zfs recv -o encryption=aes-256-gcm -o keyformat=passphrase -o keylocation=file:///testfile -o dedup=verify -o recordsize=1M -o compression=zstd-10 -o atime=off -o relatime=on main/test2`

**Reply by slavanap** [reply]
but as I've noted from previous answer, recordsize does really nothing on receive, so can be omitted for backup storage, but kept intact for sending size.

**Reply by rincebrain** [reply]
It applies for future writes, though if this is archive-only storage, that only matters if you ever restore it, potentially.

`dedup=verify` is also by far the slowest possible combination - doing sha256, and then a bytewise comparison anyway, and only deduplicates within the same encryptionroot anyway.

If you're going to use dedup (and you shouldn't, in almost every case), and you're going to use verify (same statement), you should probably opt for the fastest thing it'd let you use `verify` with, which would be `edonr`. Failing that, `skein` (2.1) or `blake3` (2.2) would be the fastest thing you don't have to use `verify` with.

(Of course, if you change this, things after you change it won't deduplicate against things before, so you'd probably want to re-receive the backup entirely...)

**Reply by slavanap** [reply]
That's good to know about `edonr`. Thanks! My primary goal is to have the most compact cold storage of about 18TB worth of data accumulated for 3 years, which could have potential hash collisions in files, has same recordsize, and occurrences of file copying within 1 dataset.

**Reply by rincebrain** [reply]
If you're accidentally colliding a salted nontrivial hash, I have questions, and I suspect various TLAs have job proposals.
